control-plane/pkg/kafka/consumer_group_lag.go

```diff
@@ -0,0 +1,228 @@
+		saramaOffsetStrategy = defaultOffsetStrategy
+	}
+	return &consumerGroupLagProvider{client: client, adminFunc: adminFunc, offsetStrategy: saramaOffsetStrategy}
+}
+
+// GetLag returns consumer group lag for a given group.
+func (p *consumerGroupLagProvider) GetLag(topic, consumerGroup string) (ConsumerGroupLag, error) {
+
+	admin, err := p.adminFunc(p.client)
+	if err != nil {
+		return ConsumerGroupLag{}, fmt.Errorf("failed to create admin client: %w", err)
+	}
+	// Note: Do not close admin since it closes the underlying client.
+
+	partitions, err := getPartitionsForTopic(admin, topic)
+	if err != nil {
+		return ConsumerGroupLag{}, err
+	}
+
+	offsets, err := admin.ListConsumerGroupOffsets(consumerGroup, map[string][]int32{topic: partitions})
+	if err != nil {
+		return ConsumerGroupLag{}, fmt.Errorf("failed to list consumer group offsets: %w", err)
+	}
+
+	consumerGroupLag := ConsumerGroupLag{
+		Topic:         topic,
+		ConsumerGroup: consumerGroup,
+		ByPartition:   make([]PartitionLag, len(partitions)),
+	}
+	for _, partition := range partitions {
+		partitionLag, err := p.getPartitionLag(partition, topic, offsets)
+		if err != nil {
+			return consumerGroupLag, fmt.Errorf("failed to get lag for partition %d: %w", partition, err)
+		}
+		consumerGroupLag.ByPartition[partition] = partitionLag
+	}
+	return consumerGroupLag, nil
+}
+
+func getPartitionsForTopic(admin sarama.ClusterAdmin, topic string) ([]int32, error) {
+
+	topicsMetadata, err := admin.DescribeTopics([]string{topic})
+	if err != nil {
+		return nil, fmt.Errorf("failed to describe topic %s: %w", topic, err)
+	}
+	if len(topicsMetadata) != 1 {
+		return nil, fmt.Errorf("unexpected number of topic metadata for topic %s: %d", topic, len(topicsMetadata))
+	}
+
+	partitionMetadata := topicsMetadata[0].Partitions
+	partitions := make([]int32, 0, len(partitionMetadata))
+	for _, p := range partitionMetadata {
+		partitions = append(partitions, p.ID)
+	}
+	return partitions, nil
+}
+
+func (p *consumerGroupLagProvider) getPartitionLag(partition int32, topic string, offsets *sarama.OffsetFetchResponse) (PartitionLag, error) {
+
+	// Get the offset of the message that will be consumed next.
+	block := offsets.GetBlock(topic, partition)
+	if block == nil {
+		return PartitionLag{}, fmt.Errorf("failed to find offset block for partition %d of topic %s", partition, topic)
+	}
+	consumerOffset := block.Offset
+
+	// Get the offset of the message that will be produced next.
+	latestOffset, err := p.client.GetOffset(topic, partition, sarama.OffsetNewest)
+	if err != nil {
+		return PartitionLag{}, fmt.Errorf("failed to find latest offset for topic %s and partition %d", topic, partition)
+	}
+	latestOffset = max(0, latestOffset) // latest offset should always be greater or equal to 0.
+
+	if consumerOffset <= invalidOffset {
+		// When we receive an invalid consumer offset, it means no offset has yet been committed.
+
+		if p.offsetStrategy == sarama.OffsetOldest {
```

Comment:
This offset strategy thing is weird... I don't see why I should provide it to default some value. If no offset is committed, the lag is just 0 and the consumerOffset is -1, right?

Reply:
No committed offset doesn't mean that the lag is 0 because the producer might have produced messages and depending on the strategy it could be 0 (OffsetNewest) or N (OffsetOldest).

Reply:
Then can we just represent the state "no offset committed"? Otherwise, this information get lost, which might be useful otherwise (e.g. I'm thinking to unordered delivery tests)

Reply:
I think propagating "no offset committed" to the caller is a good idea but I feel the logic to determine the consumer offset naturally belongs to this function.

I'd add a field called `OffsetCommitted` to determine whether an offset has been committed.

WDYT?

Reply:
Ok

Reply:
Done